File: C++/0131-palindrome-partitioning.cpp
class Solution {
private:
  void dfs(const string& s, int index, vector<string>& path, vector<vector<string>>& res){
    if(index == s.size()){
      res.push_back(path);
      return;
    }
    for(int i = index; i < s.size(); i++){
      if(isPalindrome(s,index,i)){
        path.push_back(s.substr(index, i - index + 1));
        dfs(s, i + 1, path, res);
        path.pop_back();
      }
    }
  }
  bool isPalindrome(const string& s, int l, int r){
    while(l < r)
      if(s[l++] != s[r--])
        return false;
    return true;
  }
public:
    vector<vector<string>> partition(string s) {
       vector<vector<string>> res;
       vector<string> path;
       dfs(s, 0, path, res);
       return res;
    }
};

File: Python/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
       res = []
       path = []
       self.dfs(s, 0, path, res)
       return res
    def dfs(self, s: str, index: int, path: List[str], res: List[List[str]]):
        if index == len(s):
            res.append(path)
            return
        for i in range(index, len(s)):
            if self.isPalindrome(s[index:i+1]):
                self.dfs(s, i+1, path + [s[index:i+1]], res)
    def isPalindrome(self, s: str) -> bool:
        for i in range(len(s) // 2):
            if(s[i] != s[len(s) -1 - i]):
                return False
        return True